### DISCUSSION TITLE
Return not found or internal error

### DISCUSSION THREAD

**Discussion by ivnsch** [rwf2/Rocket#2155] [Questions]
I want to return a 404 if the item isn't present in the database, and for other things just use the default (which from what I read in the docs is 500).

Something roughly like this:

```rust
#[get("/image/<id>", format = "binary")]
fn get_image(dao: State<Mutex<Box<dyn ImageDao>>>, id: String) -> Result<Vec<u8>, SomeGenericError> {
    let dao = dao.lock()?;
    match dao.load_image(&id)? {
        Some(bytes) => Ok(bytes),
        None => Err(StatusCode::NotFound),
    }
}
```

Been looking at the docs and examples but can't find anything non-basic? Also, custom responder feels like overkill for this common case?

**Comment by blackghost1987** [comment]
You don't need a full-blown custom Responder for this, because `(Status, T)` already implements Responder, see:
https://api.rocket.rs/v0.5-rc/rocket/response/trait.Responder.html#impl-Responder%3C%27r%2C%20%27o%3E-for-(Status%2C%20R)

So the only thing you need is a conversion from your Error type to (Status, Error). You can do this with a convert::From implementation which would match the error and attach a Status code, something like this (not tested code, just guidelines):
```
pub type ErrorResponse = (Status, Error); // Or maybe (Status, Json<Error>)

impl convert::From<Error> for ErrorResponse {
    fn from(e: Error) -> ErrorResponse {
      match e {
         NotFoundError => (Status::NotFound, e),
         OtherError => (Status::InternalServerError, e),
      }
   }
}
```

Then you can convert any error with `.map_err(ErrorResponse::from)` or you can simply use `?`, which should automatically call it as long as it's only one level of conversion.

**Comment by ivnsch** [comment]
Thanks for the quick reply. I'm confused, this still requires my error (`T` / `Error`) to implement `Responder`?

This is my current code:

```rust
#[get("/image/<id>", format = "binary")]
fn get_image(
    dao: State<Mutex<Box<dyn ImageDao>>>,
    id: String,
) -> Result<Vec<u8>, (Status, anyhow::Error)> {
    let dao = dao
        .lock()
        .map_err(|_| (Status::InternalServerError, anyhow!("...".to_owned())))?;

    let bytes = dao
        .load_image(&id)
        .map_err(|_| (Status::InternalServerError, anyhow!("...".to_owned())))?;

    match bytes {
        Some(bytes) => Ok(bytes),
        None => Err((Status::NotFound, anyhow!("...".to_owned()))),
    }
}
```
This is the output:

> Error: Response was a non-`Responder` `Err`: (Status { code: 404, reason: "Not Found" }, ...).
> Warning: This `Responder` implementation has been deprecated.
> Warning: In Rocket v0.5, `Result<T, E>` implements `Responder` only if `E` implements `Responder`. For the previous behavior, use `Result<T, Debug<E>>` where `Debug` is `rocket::response::Debug`.

(And if I address the deprecated warning and return `rocket::response::Debug<Error>` or `rocket::response::Debug<(Status, Error)>`, it returns always 500).

**Comment by ivnsch** [comment]
I also just tried using `std::io::Error` as the Error, which according to the [docs](https://api.rocket.rs/v0.5-rc/rocket/response/trait.Responder.html#impl-Responder%3C%27r%2C%20%27static%3E-for-Error) implements `Responder`, and still get the same error?

**Comment by blackghost1987** [comment]
An even easier solution I found for returning 404 in specific cases is just using `Result<Option<T>, Error>` for your return type. Make sure your database function won't give an error on missing rows, or at least convert them to None manually. (Diesel has an `.optional()` function for this) This way any "real" errors will go through the default 500 catcher, while the None case will return a 404.

But putting the status codes aside, if you want to return every possible Error in your response, then your Error type has to implement Responder. If you don't care about having a structured error, then an idea I've seen before is to use the Anyhow crate and convert every error to String, and return the plain text error message (or wrap it in a Json).

I don't have any experience with `rocket::response::Debug`, but it looks like you should probably use `(Status, rocket::response::Debug<Error>)`, because if the Debug is on the outside, it will probably debug print the Status code instead of using it as a response. But this might not be enough still, the documentation says: "Debug prints the internal value before forwarding to the 500 error catcher." So Debug might always respond with 500 whatever you do, but it might be still useful combined with the "Option" solution above.

I'm not sure why `std::io::Error` gives you a problem, `(Status, std::io::Error)` definitely implements Responder and should work. Maybe an import issue?

**Comment by ivnsch** [comment]
Ooh, returning `Option` worked (which intuitively makes sense) - thanks!

For completeness - about the other issue, I realized that I was using `0.4.10`, probably because the `0.5.0-rc.1` badge in the readme links to it (because 0.5.0 isn't a release version?)
While now `std::io::Error` is a `Responder`, it ignores `Status` and just returns 500:
> I/O Error: Custom { kind: Other, error: "not found!" }
> Outcome: Failure
> No 500 catcher registered. Using Rocket default.

This is the code for that:
```rust
#[get("/image/<id>", format = "binary")]
fn get_image(
    dao: &State<Box<dyn ImageDao>>,
    id: String,
) -> Result<Vec<u8>, (Status, std::io::Error)> {
    let bytes = dao.load_image(&id).map_err(|e| {
        (
            Status::InternalServerError,
            std::io::Error::new(ErrorKind::Other, format!("oh no {e:?}!")),
        )
    })?;

    match bytes {
        Some(bytes) => Ok(bytes),
        None => Err((
            Status::NotFound,
            std::io::Error::new(ErrorKind::Other, "not found!"),
        )),
    }
}
```

Would like to also ask the maintainers to consider adding examples about this / a more comprehensive error handling guide? I found 2 issues asking similar questions (https://github.com/SergioBenitez/Rocket/issues/71, https://github.com/SergioBenitez/Rocket/issues/72) which, in addition to this thread (which also has different solutions), seems indicative of a documentation deficiency. This should be a common use case. I just started using Rocket and while I appreciate the responsiveness here, it has been a bit frustrating.
